flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPoolTest.java

```diff
@@ -518,4 +527,204 @@ public void go() throws Exception {
+	 * to the global network buffer pool.
+	 */
+	@Test(timeout = 10000L)
+	public void testBlockingRequestFromMultiLocalBufferPool() throws Exception {
+		final int localPoolRequiredSize = 5;
+		final int localPoolMaxSize = 10;
+		final int numLocalBufferPool = 2;
+		final int numberOfSegmentsToRequest = 10;
+		final int numBuffers = numLocalBufferPool * localPoolMaxSize;
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(numLocalBufferPool);
+		final NetworkBufferPool globalPool = new NetworkBufferPool(numBuffers, 128, numberOfSegmentsToRequest);
+		final List<BufferPool> localBufferPools = new ArrayList<>(numLocalBufferPool);
+
+		try {
+			// create local buffer pools
+			for (int i  = 0; i < numLocalBufferPool; ++i) {
+				final BufferPool localPool = globalPool.createBufferPool(localPoolRequiredSize, localPoolMaxSize);
+				localBufferPools.add(localPool);
+				assertTrue(localPool.isAvailable().isDone());
+			}
+
+			// request some segments from the global pool in two different ways
+			final List<MemorySegment> segments = new ArrayList<>(numberOfSegmentsToRequest - 1);
+			for (int i = 0; i < numberOfSegmentsToRequest - 1; ++i) {
+				segments.add(globalPool.requestMemorySegment());
+			}
+			final List<MemorySegment> exclusiveSegments = globalPool.requestMemorySegments();
+			assertTrue(globalPool.isAvailable().isDone());
+			for (BufferPool localPool: localBufferPools) {
+				assertTrue(localPool.isAvailable().isDone());
+			}
+
+			// blocking request buffers form local buffer pools
+			final CountDownLatch latch = new CountDownLatch(numLocalBufferPool);
+			final BlockingQueue<BufferBuilder> segmentsRequested = new ArrayBlockingQueue<>(numBuffers);
+			for (final BufferPool localPool: localBufferPools) {
+				executorService.submit(() -> {
+					for (int num = localPoolMaxSize; num > 0; --num) {
+						ExceptionUtils.suppressExceptions(
+							() -> segmentsRequested.add(localPool.requestBufferBuilderBlocking()));
+					}
+					latch.countDown();
+				});
+			}
+
+			// wait until all available buffers are requested
+			while (globalPool.getNumberOfAvailableMemorySegments() > 0) {
+				Thread.sleep(100);
+			}
+
+			final CompletableFuture<?> globalPoolAvailableFuture = globalPool.isAvailable();
+			assertFalse(globalPoolAvailableFuture.isDone());
+
+			final List<CompletableFuture<?>> localPoolAvailableFutures = new ArrayList<>(numLocalBufferPool);
+			for (BufferPool localPool: localBufferPools) {
+				CompletableFuture<?> localPoolAvailableFuture = localPool.isAvailable();
+				localPoolAvailableFutures.add(localPoolAvailableFuture);
+				assertFalse(localPool.isAvailable().isDone());
+			}
+
+			// recycle the previously requested segments
+			for (MemorySegment segment: segments) {
+				globalPool.recycle(segment);
+			}
+			globalPool.recycleMemorySegments(exclusiveSegments);
+
+			assertTrue(globalPoolAvailableFuture.isDone());
```

Comment:
This verify is not stable strictly, because if the above recycled segments are requested away by above threads, then it would be unavailable again? Or we could adjust `numBuffers = numLocalBufferPool * localPoolMaxSize + 1` to make the global pool must available here.

Reply:
The future should be always completed even the recycled segments are taken away.